### PR TITLE
Audio HIP PR7 - Spectrogram HIP Support

### PR DESCRIPTION
**Pull request by sampath1117**
- Adds support for Spectrogram in HIP Backend
- Adds relevant qa tests and performance test support

### CODE REVIEW

**Comment by sampath1117** [comment]
@r-abishek 
Added the changes discussed in morning call

- Added in documentation about windowFn 
- Added in header file w.r.t max size required for spectrogram and done validation checks
- Added in main readme about list of audio functions supported
- Moved constant compute of numTiles, outside the kernel
- Modified the comments for index checking to make it more clear for fft kernel
- Used srcIdx, dstIdx for the coefficient compute kernel
- Removed t * hipBlockDim_x and used a loop counter 
- Added comment for vertical in fft kernel

For adding spectrogram output image
- Actually spectrogram jupyter notebook is not yet merged in rocAL TOT

It is issued to opensource but not yet approved and merged
https://github.com/ROCm/rocAL/pull/181

So i have not added the spectrogram output image in readme for now

**Comment by sampath1117** [comment]
I have also incremented the MINOR_VERSION_NUMBER in rpp
https://github.com/sampath1117/rpp/blob/sr/spectrogram_hip/include/rpp_version.h#L42

Lakshmi said, we need to increment this version number for every new feature we add in RPP
For audio, she mentioned once all audio PR's are merged we can increment the version number

File: utilities/test_suite/HIP/Tensor_audio_hip.cpp [hunk 1]
```diff
@@ -116,6 +116,11 @@ int main(int argc, char **argv)
     iBufferSize = (Rpp64u)srcDescPtr->h * (Rpp64u)srcDescPtr->w * (Rpp64u)srcDescPtr->c * (Rpp64u)srcDescPtr->n;
     oBufferSize = (Rpp64u)dstDescPtr->h * (Rpp64u)dstDescPtr->w * (Rpp64u)dstDescPtr->c * (Rpp64u)dstDescPtr->n;
 
+    // compute maximum possible buffer size of spectrogram
+    unsigned long long spectrogramMaxBufferSize = 257 * 3754 * dstDescPtr->n;
```
Comment (r-abishek) [inline utilities/test_suite/HIP/Tensor_audio_hip.cpp:120]: Can we use some RPP type?

File: utilities/test_suite/HIP/Tensor_audio_hip.cpp [hunk 2]
```diff
@@ -236,6 +240,62 @@ int main(int argc, char **argv)
...
+                    // check if the output buffer size is greater than predefined spectrogramMaxBufferSize
+                    if (dstDescPtr->n * dstDescPtr->strides.nStride > spectrogramMaxBufferSize)
+                    {
+                        std::cout << "\nError! Requested spectrogram output size is greater than predefined max size for spectrogram in test suite."
+                                     "\nPlease modify spectrogramMaxBufferSize value in test suite for running spectrogram kernel" << std::endl;
+                        exit(0);
+                    }
+
+                    startWallTime = omp_get_wtime();
+                    rppt_spectrogram_gpu(d_inputf32, srcDescPtr, d_outputf32, dstDescPtr, srcLengthTensor, centerWindows, reflectPadding, windowFn, nfft, power, windowLength, windowStep, handle);
```
Comment (r-abishek) [inline utilities/test_suite/HIP/Tensor_audio_hip.cpp:295]: Please put the whole spectrogram setup into a function like init_ricap() in the test suite header

Comment (r-abishek) [reply utilities/test_suite/HIP/Tensor_audio_hip.cpp:295]: Same comment if host setup is also long, and you can take common code out

File: src/modules/rppt_tensor_audio_augmentations.cpp [hunk 3]
```diff
@@ -438,5 +438,58 @@ RppStatus rppt_pre_emphasis_filter_gpu(RppPtr_t srcPtr,
...
+                               bool reflectPadding,
+                               Rpp32f *windowFunction,
+                               Rpp32s nfft,
+                               Rpp32s power,
+                               Rpp32s windowLength,
+                               Rpp32s windowStep,
+                               rppHandle_t rppHandle)
+{
+#ifdef HIP_COMPILE
+    if ((dstDescPtr->layout != RpptLayout::NFT) && (dstDescPtr->layout != RpptLayout::NTF)) return RPP_ERROR_INVALID_DST_LAYOUT;
```
Comment (r-abishek) [inline src/modules/rppt_tensor_audio_augmentations.cpp:458]: Put the return on next line like couple lines below

File: src/modules/hip/kernel/spectrogram.hpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    sinFactor_smem[hipThreadIdx_y][hipThreadIdx_x] = 0.0f;  // initialization of shared memory to 0 using all 16 x 16 threads
+    __syncthreads();
+
+    int srcIdx = id_z * srcStrideNH.x + id_y * srcStrideNH.y;
+    float realVal = 0.0f, imaginaryVal = 0.0f;
+    for(int t = 0, offset = 0; t < numTiles; t++, offset += hipBlockDim_x)
+    {
+        // load input values to shared memory if (id_y, srcCol) < (numWindows, nfft) - range of input
+        int srcCol = (offset + hipThreadIdx_x);
+        int factorRow = (offset  + hipThreadIdx_y);
```
Comment (r-abishek) [inline src/modules/hip/kernel/spectrogram.hpp:106]: double space

File: src/modules/hip/kernel/spectrogram.hpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        for (int j = 0; j < hipBlockDim_x; j++)
+        {
+            realVal += (input_smem[hipThreadIdx_y][j] * cosFactor_smem[j][hipThreadIdx_x]);
+            imaginaryVal += (-input_smem[hipThreadIdx_y][j] * sinFactor_smem[j][hipThreadIdx_x]);
+        }
+        __syncthreads();
+    }
+
+    // final store to dst
+    if (id_y < numWindows && id_x < numBins)
```
Comment (r-abishek) [inline src/modules/hip/kernel/spectrogram.hpp:131]: Add parentheses for multiple checks in if

File: src/modules/hip/kernel/spectrogram.hpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    int srcLength = srcLengthTensor[id_z];
+    int numWindows = numWindowsTensor[id_z];
+    int nfft = params_i4.x;
+    int windowLength = params_i4.y;
+    int windowStep = params_i4.z;
+    int windowCenterOffset = params_i4.w;
+
+    if (id_x >= windowLength || id_y >= numWindows)
```
Comment (r-abishek) [inline src/modules/hip/kernel/spectrogram.hpp:27]: Same comment for all instances in this and function below too

File: src/modules/hip/kernel/spectrogram.hpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+                                                float *sinFactor,
+                                                int numBins,
+                                                int nfft)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    if (id_y >= numBins || id_x >= nfft)
+        return;
+
+    float factor = id_x * ((2.0f * id_y * M_PI) / nfft);
```
Comment (r-abishek) [inline src/modules/hip/kernel/spectrogram.hpp:63]: Can we have a #define for 2 * PI?

**Review by r-abishek — COMMENTED** [review]
Added some more comments

File: CMakeLists.txt [hunk 8]
```diff
@@ -236,6 +236,14 @@ if("${BACKEND}" STREQUAL "HIP")
     message("-- ${White}${PROJECT_NAME} -- AMDGPU_TARGETS: ${AMDGPU_TARGETS}${ColourReset}")
 
     find_package(HIP QUIET)
+    find_package(HIPFFT REQUIRED)
+    if (HIPFFT_FOUND)
```
[the remark this answers is not in the document]
Comment (r-abishek) [reply CMakeLists.txt]: ignore if removed

File: CMakeLists.txt [hunk 9]
```diff
@@ -249,6 +257,8 @@ if("${BACKEND}" STREQUAL "HIP")
         link_directories(${HIP_PATH}/${CMAKE_INSTALL_LIBDIR})
         # link To HIP Host library -- [hip::host] to use host API
         set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} hip::host)
+        set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} hipfft)
+        set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} rocfft)
```
[the remark this answers is not in the document]
Comment (r-abishek) [reply CMakeLists.txt]: ignore if removed

**Pull request merged**